examples/virus_antibody/agents.py

```diff
@@ -0,0 +1,259 @@
+"""
+Mesa implementation of Virus/Antibody model: Agents module.
+"""
+
+import copy
+import os
+import sys
+import weakref
+from collections import deque
+
+import numpy as np
+
+sys.path.insert(0, os.path.abspath("../../mesa"))
+from mesa.experimental.continuous_space import ContinuousSpaceAgent
+
+
+class AntibodyAgent(ContinuousSpaceAgent):
+    """An Antibody agent. They move randomly until they see a virus, go fight it.
+    If they lose, stay KO for a bit, lose health and back to random moving.
+    """
+
+    def __init__(
+        self,
+        model,
+        space,
+        sight_range,
+        duplication_rate,
+        ko_timeout,
+        memory_capacity,
+        initial_position=(0, 0),
+        direction=(1, 1),
+    ):
+        super().__init__(model=model, space=space)
+
+        # Movement & state
+        self.position = initial_position
+        self.speed = 1.5
+        self.direction = np.array(direction, dtype=float)
+
+        # Characteristics
+        self.sight_range = sight_range
+        self.health = 2
+        self.duplication_rate = duplication_rate
+
+        # Memory
+        self.st_memory: deque = deque()
+        self.lt_memory: list = []
+        self.memory_capacity = memory_capacity
+
+        # Target & KO state
+        self.target = None  # will hold a weakref.ref or None
+        self.ko_timeout = ko_timeout
+        self.ko_steps_left = 0
+
+    def step(self):
+        # Acquire a virus target if we don't already have one
+        if self.target is None:
+            closest = self.find_closest_virus()
+            if closest:
+                self.target = weakref.ref(closest)
+
+        # Communicate and maybe duplicate
+        self.communicate()
+        if self.random.random() < self.duplication_rate:
+            self.duplicate()
+
+        # Then move
+        self.move()
+
+    def find_closest_virus(self):
+        agents, _ = self.space.get_agents_in_radius(self.position, self.sight_range)
+        viruses = [a for a in agents if isinstance(a, VirusAgent)]
+        return viruses[0] if viruses else None
+
+    def communicate(self) -> bool:
+        agents, _ = self.space.get_agents_in_radius(self.position, self.sight_range)
```

Comment:
You do the `space.get_agents_in radius` twice. Once in find_closest_virus, and once in communicate. It would be computationally more efficient to do it once in step, split the return in virus and antibody agents, and use these respectively in the helper methods. This will substantially reduce runtime.